pkg/service/blobs/server.go

```diff
@@ -0,0 +1,144 @@
+	blobs     blobstore.Blobstore
+	presigner presigner.RequestPresigner
+	allocs    allocationstore.AllocationStore
+}
+
+func NewServer(presigner presigner.RequestPresigner, allocs allocationstore.AllocationStore, blobs blobstore.Blobstore) (*Server, error) {
+	return &Server{blobs, presigner, allocs}, nil
+}
+
+func (srv *Server) Serve(mux *http.ServeMux) {
+	mux.HandleFunc("GET /blob/{blob}", NewBlobGetHandler(srv.blobs))
+	mux.HandleFunc("PUT /blob/{blob}", NewBlobPutHandler(srv.presigner, srv.allocs, srv.blobs))
+}
+
+func NewBlobGetHandler(blobs blobstore.Blobstore) func(http.ResponseWriter, *http.Request) {
+	if fsblobs, ok := blobs.(blobstore.FileSystemer); ok {
+		serveHTTP := http.FileServer(fsblobs.FileSystem()).ServeHTTP
+		return func(w http.ResponseWriter, r *http.Request) {
+			_, bytes, err := multibase.Decode(r.PathValue("blob"))
+			if err != nil {
+				http.Error(w, fmt.Sprintf("decoding multibase encoded digest: %s", err), http.StatusBadRequest)
+				return
+			}
+
+			digest, err := multihash.Cast(bytes)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("invalid multihash digest: %s", err), http.StatusBadRequest)
+				return
+			}
+
+			r.URL.Path = fsblobs.EncodePath(digest)
+			serveHTTP(w, r)
+		}
+	}
+
+	log.Error("blobstore does not support filesystem access")
+	return func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "not supported", http.StatusInternalServerError)
+	}
+}
+
+func NewBlobPutHandler(presigner presigner.RequestPresigner, allocs allocationstore.AllocationStore, blobs blobstore.Blobstore) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		_, sHeaders, err := presigner.VerifyUploadURL(r.Context(), *r.URL, r.Header)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusUnauthorized)
+			return
+		}
+
+		_, bytes, err := multibase.Decode(r.PathValue("blob"))
+		if err != nil {
+			http.Error(w, fmt.Sprintf("decoding multibase encoded digest: %s", err), http.StatusBadRequest)
+			return
+		}
+
+		digest, err := multihash.Cast(bytes)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("invalid multihash digest: %s", err), http.StatusBadRequest)
+			return
+		}
+
+		results, err := allocs.List(r.Context(), digest)
+		if err != nil {
+			log.Errorf("listing allocations: %w", err)
+			http.Error(w, "list allocations failed", http.StatusInternalServerError)
+			return
+		}
+
+		if len(results) == 0 {
+			log.Warnf("missing allocation for write to: z%s", digest.B58String())
+			http.Error(w, "missing allocation", http.StatusForbidden)
+			return
+		}
+
+		expired := true
+		for _, a := range results {
+			exp := a.Expires
+			if exp > uint64(time.Now().Unix()) {
+				expired = false
+				break
+			}
+		}
+		if expired {
+			http.Error(w, "expired allocation", http.StatusForbidden)
+			return
+		}
+
+		log.Infof("found %d allocations for write to: z%s", len(results), digest.B58String())
+
+		_, err = blobs.Get(r.Context(), digest)
+		if err == nil {
```

Comment:
Good question, currently we use `size: 0` to indicate a blob _has not_ been allocated in your space because it was already allocated and you don't need to upload... but we don't have a mechanism to say that it _has_ been allocated in your space, and you don't need to upload.

Previously I think we just let S3 handle this - allowing the data to be overwritten.

In the mean time we should probably upload to a temporary location and then move into place? Handling the move error and doing nothing if the move failed because it already exists.

Reply:
https://github.com/storacha/storage/issues/6